### PR TITLE
Product Collection: Add `Attributes` filter control to sidebar settings

### LINKED ISSUE #9363: Product Collection - Filters: Attributes
### Description
Provide "Attributes" filter to the Product Collection block. This setting exists in Products block, so should be reused if possible.

### Goals - acceptance criteria
- Changing the filter should impact the query and the change should be reflected in the Editor's preview and frontend
- Filter should be placed in Filters section

### Image
image

image

### PR DESCRIPTION
**Pull request by imanish003**
This PR adds `Attributes` filter control to sidebar settings.
image

Fixes #9363 

### Testing

1. Create a new post
2. Add Product Collection block
3. With the `Product Collection` block selected, inspect the sidebar for the `Product attributes` control. Confirm its appearance and functionalities:
   - The control should be displayed as shown in the images below:
image
image

4. Test the `Product attributes` control by selecting any attributes. For example, color -> blue:
   - After the selection, verify that changes are immediately visible in the Editor view.

5. Save and publish the post:
   - Click on `Publish...` at the top right corner.
   - After publishing, navigate to the Frontend and verify that the changes you made in the Editor are accurately reflected.

6. Return to the Editor and validate the `Reset all` functionality of the filters:
   - Make sure that clicking the `Reset all` button resets your selected filters.
   - The `Reset all` button can be found as shown in the image below:
image

7. Test the `Manage attributes` functionality:
   - Click on the `Manage attributes` link.
   - Verify that this redirects you to the designated page where you can manage product attributes

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add `Attributes` filter control to sidebar settings

### CODE REVIEW

**Comment by thealexandrelara** [comment]
Hey @imanish003,
The Product Attributes filter is not being displayed for me, is there any requirement or step that I should do to enable it? 

image

**Comment by imanish003** [comment]
Hey @thealexandrelara,

> The Product Attributes filter is not being displayed for me, is there any requirement or step that I should do to enable it?

That's strange. I can't think of any specific requirement or step to enable it. Are you using a local build? If so, you could try stopping the server and then running `npm start` again.

I've tested it on my end, and it's working fine. Here's a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/aaca8f7c-b8ad-47bd-8ecb-1e097522d638

**Comment by thealexandrelara** [comment]
> That's strange. I can't think of any specific requirement or step to enable it. Are you using a local build? If so, you could try stopping the server and then running npm start again.

Yes, it is a local build. I tried start the server again, disabling Gutenberg, running in Incognito mode but the filter is still not showing to me. Maybe we should ask someone else to also review this PR, so we can confirm if it is only something happening on my end

**Comment by imanish003** [comment]
@thealexandrelara 

> Yes, it is a local build. I tried start the server again, disabling Gutenberg, running in Incognito mode but the filter is still not showing to me. Maybe we should ask someone else to also review this PR, so we can confirm if it is only something happening on my end

Yes, that sounds like a good idea. I am requesting a review from @Aljullu. 🙂 

@Aljullu, FYI, it seems that @thealexandrelara can't see the `Attributes` filter control. Can you please review and check if you are able to see it? 🙂

File: assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	__experimentalToolsPanelItem as ToolsPanelItem,
+} from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { ProductCollectionQuery } from '../types';
+
+const EDIT_ATTRIBUTES_URL =
+	'/wp-admin/edit.php?post_type=product&page=product_attributes';
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx]: I think this link won't work if the site is hosted in a subfolder. You can use `ADMIN_URL` constant to prepend the correct path like [here](https://github.com/woocommerce/woocommerce-blocks/blob/ac2b6d6057d2e26b0ac20edba1dcdb5e8530213c/assets/js/blocks/checkout/inner-blocks/checkout-shipping-methods-block/edit.tsx#L118).

File: assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+	useEffect( () => {
+		if ( woocommerceAttributes ) {
+			setSelected(
+				woocommerceAttributes.map( ( { termId: id } ) => ( {
+					id,
+				} ) )
+			);
+		}
+	}, [ woocommerceAttributes ] );
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx]: Do you think we could convert this component to a fully controlled component so it doesn't need to hold its state? Something along these lines:

```suggestion
	const selected = woocommerceAttributes.map( ( { termId: id } ) => ( {
		id,
	} ) );
```

File: assets/js/blocks/product-collection/inspector-controls/index.tsx [hunk 3]
```diff
@@ -50,6 +58,14 @@ const ProductCollectionInspectorControls = (
 				<OnSaleControl { ...props } />
 				<StockStatusControl { ...props } />
 				<KeywordControl { ...props } />
+				<AttributesControl
+					woocommerceAttributes={
+						props.attributes.query &&
+						( props.attributes.query
+							.woocommerceAttributes as AttributeMetadata[] )
+					}
+					setQueryAttribute={ setQueryAttributeBind }
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/index.tsx:67]: I'm not sure to understand `setQueryAttributeBind`, is it equivalent to this?

```suggestion
					setQueryAttribute={ ( newQuery ) => {
						setQueryAttribute( props, newQuery );
					} }
```

**Review by Aljullu — COMMENTED** [review]
> @Aljullu, FYI, it seems that @thealexandrelara can't see the `Attributes` filter control. Can you please review and check if you are able to see it? 🙂

It's showing up as expected for me. @thealexandrelara did you check if there are any errors in the console?

@imanish003 I did a quick review and overall it's looking good, but I noticed the `Reset all` button doesn't reset the attributes, I think that's because `woocommerceAttributes` is not defined in `DEFAULT_FILTERS`.

Besides that, I left some minor comments below.

**Comment by thealexandrelara** [comment]
> It's showing up as expected for me. @thealexandrelara did you check if there are any errors in the console?

I checked the console and the only error that is showing to me is this one:
image

This error was appearing to me in other branches as well, however it was not preventing me from using other blocks. I really don't know what is happening on my side, Maybe I'm going to have to do a fresh install on my testing environment

File: assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx (hunk 1 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx]: That's a good suggestion. Utilizing the ADMIN_URL constant ensures the link works regardless of the site's hosting location in a subfolder. I made the changes in [f9a6813](https://github.com/woocommerce/woocommerce-blocks/pull/9600/commits/f9a6813300b2b9a3aa89f170f7339a4e4efaddc8).

File: assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx (hunk 2 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx]: That's a valid idea. By converting the component to a fully controlled component and using the mapped value of woocommerceAttributes for selected, we can eliminate the need for the useState hook. 🤝 

I made the changes in [f9a6813](https://github.com/woocommerce/woocommerce-blocks/pull/9600/commits/f9a6813300b2b9a3aa89f170f7339a4e4efaddc8).

File: assets/js/blocks/product-collection/inspector-controls/index.tsx (hunk 3 above)
Comment (imanish003) [reply assets/js/blocks/product-collection/inspector-controls/index.tsx:67]: Yes, You are correct. Do you have any concerns regarding the approach? 🙂

**Comment by imanish003** [comment]
Hey @Aljullu,

> @imanish003 I did a quick review and overall it's looking good, but I noticed the Reset all button doesn't reset the attributes, I think that's because woocommerceAttributes is not defined in DEFAULT_FILTERS.

Oh, Yeah. You are right. Good catch 🙌🏻 I have made the improvements/changes in [f9a6813](https://github.com/woocommerce/woocommerce-blocks/pull/9600/commits/f9a6813300b2b9a3aa89f170f7339a4e4efaddc8). Can you please take a look again? 🙂

**Comment by imanish003** [comment]
Hey @thealexandrelara 👋 

> This error was appearing to me in other branches as well, however it was not preventing me from using other blocks. I really don't know what is happening on my side. Maybe I'm going to have to do a fresh install on my testing environment.

Yeah, that error shouldn't be a problem. It makes sense to do a fresh install. Apart from this, I also try the following things in such cases:

Fix 1 - Removing the `build` directory and then running `npm start`. 
Fix 2 - Trying to run `npm install` and `composer install` again.
Fix 3 - Restarting the Local server.
Fix 4 - Making sure that the `wp-config.php` file has `define('JETPACK_AUTOLOAD_DEV', true);`.

File: assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+					setQueryAttribute( {
+						woocommerceAttributes: newValue,
+					} );
+				} }
+				operator={ 'any' }
+				isCompact={ true }
+				type={ 'token' }
+			/>
+			<ExternalLink
+				className="wc-product-collection-panel__external-link"
```
Comment (Aljullu) [inline assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx]: This class should have the [`wc-block-editor` prefix](https://github.com/woocommerce/woocommerce-blocks/blob/90760e4be19c5e719d8ae20fcb372d01ce63c191/docs/contributors/coding-guidelines.md?plain=1#LL29C6-L29C21).

File: assets/js/blocks/product-collection/inspector-controls/index.tsx (hunk 3 above)
Comment (Aljullu) [reply assets/js/blocks/product-collection/inspector-controls/index.tsx:67]: Not really, just a personal preference as I find `bind` a bit difficult to understand. :smile:

**Review by Aljullu — APPROVED** [review]
Thanks for the updates, @imanish003! This is looking great on my end. I left one minor comment, but pre-approving.

**Pull request merged**